manager/container.go

```diff
@@ -283,106 +285,127 @@ func (cd *containerData) getContainerPids(inHostNamespace bool) ([]string, error
+	}
 	processes := []v2.ProcessInfo{}
 	lines := strings.Split(string(out), "\n")
 	for _, line := range lines[1:] {
-		if len(line) == 0 {
-			continue
-		}
-		fields := strings.Fields(line)
-		if len(fields) < expectedFields {
-			return nil, fmt.Errorf("expected at least %d fields, found %d: output: %q", expectedFields, len(fields), line)
-		}
-		pid, err := strconv.Atoi(fields[1])
-		if err != nil {
-			return nil, fmt.Errorf("invalid pid %q: %v", fields[1], err)
-		}
-		ppid, err := strconv.Atoi(fields[2])
-		if err != nil {
-			return nil, fmt.Errorf("invalid ppid %q: %v", fields[2], err)
-		}
-		percentCPU, err := strconv.ParseFloat(fields[4], 32)
-		if err != nil {
-			return nil, fmt.Errorf("invalid cpu percent %q: %v", fields[4], err)
-		}
-		percentMem, err := strconv.ParseFloat(fields[5], 32)
-		if err != nil {
-			return nil, fmt.Errorf("invalid memory percent %q: %v", fields[5], err)
-		}
-		rss, err := strconv.ParseUint(fields[6], 0, 64)
+		process, err := cd.parsePsLine(line, cadvisorContainer, inHostNamespace)
 		if err != nil {
-			return nil, fmt.Errorf("invalid rss %q: %v", fields[6], err)
+			return nil, fmt.Errorf("could not parse line %s: %v", line, err)
 		}
-		// convert to bytes
-		rss *= 1024
-		vs, err := strconv.ParseUint(fields[7], 0, 64)
-		if err != nil {
-			return nil, fmt.Errorf("invalid virtual size %q: %v", fields[7], err)
-		}
-		// convert to bytes
-		vs *= 1024
-		psr, err := strconv.Atoi(fields[11])
-		if err != nil {
-			return nil, fmt.Errorf("invalid pid %q: %v", fields[1], err)
-		}
-
-		cgroup, err := cd.getCgroupPath(fields[12])
-		if err != nil {
-			return nil, fmt.Errorf("could not parse cgroup path from %q: %v", fields[11], err)
-		}
-		// Remove the ps command we just ran from cadvisor container.
-		// Not necessary, but makes the cadvisor page look cleaner.
-		if !inHostNamespace && cadvisorContainer == cgroup && fields[10] == "ps" {
+		if process == nil {
 			continue
 		}
-		var cgroupPath string
-		if isRoot {
-			cgroupPath = cgroup
-		}
 
 		var fdCount int
-		dirPath := path.Join(rootfs, "/proc", strconv.Itoa(pid), "fd")
+		dirPath := path.Join(rootfs, "/proc", strconv.Itoa(process.Pid), "fd")
 		fds, err := ioutil.ReadDir(dirPath)
 		if err != nil {
 			klog.V(4).Infof("error while listing directory %q to measure fd count: %v", dirPath, err)
 			continue
 		}
 		fdCount = len(fds)
+		process.FdCount = fdCount
 
-		if isRoot || cd.info.Name == cgroup {
-			processes = append(processes, v2.ProcessInfo{
-				User:          fields[0],
-				Pid:           pid,
-				Ppid:          ppid,
-				StartTime:     fields[3],
-				PercentCpu:    float32(percentCPU),
-				PercentMemory: float32(percentMem),
-				RSS:           rss,
-				VirtualSize:   vs,
-				Status:        fields[8],
-				RunningTime:   fields[9],
-				Cmd:           fields[10],
-				CgroupPath:    cgroupPath,
-				FdCount:       fdCount,
-				Psr:           psr,
-			})
+		if cd.isRoot() || cd.info.Name == process.CgroupPath {
```

Comment:
I think the logic here has changed slightly regarding what processInfos are ultimately added to `processes` with respect to the the `cd.info.Name == process.CgroupPath` check

Old logic checked `cd.info.Name == cgroup`
https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L364

where `cgroup` was always populated:

https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L341

However, in the change here, `process.CgroupPath` will only be populated if the containerData is root:

https://github.com/google/cadvisor/blob/f34a89b2be44fe4915e7f827a4b60dcac9007230/manager/container.go#L393-L394

I think that might be why the old code (very confusingly) had both `cgroup` and `cgroupPath` where cgroupPath was what was set on the v2.ProcessInfo.CGroupPath struct:

https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L341-L353

As a result, I think this is currently only including processInfo's that are `isRoot()==true`  which is probably skipping more processInfos than we actually want.

Reply:
I think I managed to fix the problem that you described. If you take a look at added test cases then you will see that:
* all processes (except `ps` are returned when root container if requested
* only processes belonging to requested container (with no cgroup information set) are returned when it is a non-root container

Reply:
Thanks for the fix, your change makes sense!